Objects/setobject.c

```diff
@@ -2334,6 +2334,12 @@ set_init(PySetObject *self, PyObject *args, PyObject *kwds)
     if (!PyArg_UnpackTuple(args, Py_TYPE(self)->tp_name, 0, 1, &iterable))
         return -1;
 
+    if (self->fill == 0 && Py_REFCNT(self) == 1) {
```

Comment:
I think we should do the `Py_REFCNT()` check first so that we don't access `self->fill` without a lock in case there are other threads with references modifying `self`.

Also, do we need to set `self->hash = -1` as we do below?

Reply:
> Also, do we need to set self->hash = -1 as we do below?

Well it will be more safe.